gpii/node_modules/flowManager/src/UserLogonStateChange.js

```diff
@@ -0,0 +1,221 @@
+/*
+ * GPII User Logon State Change
+ *
+ * Copyright 2012 OCAD University
+ * Copyright 2015 Raising the Floor - International
+ *
+ * Licensed under the New BSD license. You may not use this file except in
+ * compliance with this License.
+ *
+ * The research leading to these results has received funding from the European Union's
+ * Seventh Framework Programme (FP7/2007-2013)
+ * under grant agreement no. 289016.
+ *
+ * You may obtain a copy of the License at
+ * https://github.com/GPII/universal/blob/master/LICENSE.txt
+ */
+
+(function () {
+
+    "use strict";
+
+    var fluid = require("infusion");
+    var gpii = fluid.registerNamespace("gpii");
+    fluid.registerNamespace("gpii.request.flowManager.userLogonStateChange");
+
+    gpii.request.flowManager.userLogonStateChange.handleResetToken = function (activeToken, that) {
+        if (activeToken === undefined) { // if no user is logged in
+            that.errorResponse("No users currently logged in - nothing to reset");
+        } else {
+            fluid.log("Logging out the user with token " + activeToken);
+            that.logoutUser(activeToken);
+        }
+    };
+
+    /**
+     * /user/:userToken/logonChange handler
+     * * If no user is logged in, user <userToken> will be logged in
+     * * If <userToken> is logged in, <userToken> will be logged out
+     * * If another user is logged in, nothing happens
+     * If "reset" is used as the <mytoken> for any of the above URLs, any currently logged in user will be logged out
+     */
+    gpii.request.flowManager.userLogonStateChange.logonChangeHandler = function (userToken, lifecycleManager, that) {
+        // find currently logged in user:
+        var activeTokens = lifecycleManager.getActiveSessionTokens();
+        if (userToken === "reset") {
+            gpii.request.flowManager.userLogonStateChange.handleResetToken(activeTokens[0], that);
+            return;
+        }
+        if (activeTokens.length === 0) { // if no user is logged in
+            that.loginUser(userToken);
+            return;
+        }
+
+        if (activeTokens.length === 1) {
+            if (activeTokens[0] === userToken) {
+                that.logoutUser(activeTokens[0]);
+            } else {
+                that.errorResponse("Got keyaction from user " + userToken + ", but the user " +
+                    activeTokens[0] + " is already logged in. So ignoring key action.");
```

Comment:
The default 500 status code is probably inappropriate for these responses - the server has not failed, but the client has sent a request at an inappropriate time. We might consider something like 400 or 401 more appropriate - certainly something in the 4xx range